### PR TITLE
feat: add support for select, offset, startAt, startAfter, endAt, end…

### PR DESCRIPTION
**Pull request by alextekartik**
…Before, FieldValue.delete, FieldValue.timestamp

select: all selecting fields. varargs needed here so original binding does not work
offset: was simply missing
startAt, startAfter, endAt, endBefore: copy what is done in firebase-dart (web)
FieldValue.timestamp and FieldValue.delete: add support for it

add unit tests

### CODE REVIEW

File: lib/src/firestore.dart [hunk 1]
```diff
@@ -348,6 +367,41 @@ class _FirestoreData {
...
+        default:
+          throw new ArgumentError.value(
+              value, key, 'Unsupported value type for Firestore.');
+      }
+    }
+    setProperty(nativeInstance, key, data);
+  }
+
+  // only for testing we should never read such value
+  FieldValue getFieldValue(String key) {
```
Comment (pulyaevskiy) [inline lib/src/firestore.dart]: Is it possible to write an integration test and avoid having this method at all?

It seems that we'd just need to set a field value, write to the database, read the document back and verify that our field is either not there (for deleted) or contains a non-empty DateTime?

File: lib/src/firestore.dart [hunk 2]
```diff
@@ -348,6 +367,41 @@ class _FirestoreData {
     setProperty(nativeInstance, key, data);
   }
 
+  void setFieldValue(String key, FieldValue value) {
```
Comment (pulyaevskiy) [inline lib/src/firestore.dart]: Wondering if we can replace this method with two dedicated methods for each special value? E.g. `setServerTimestamp(String key)` and `deleteField(String key)`?

This would also allow to remove `fieldValueDeleted`, `fieldValueServerTimestamp` and the new `FieldValue` enum.

File: lib/src/firestore.dart [hunk 3]
```diff
@@ -681,9 +745,9 @@ class DocumentQuery {
   /// The [values] must be in order of [orderBy] filters.
   ///
   /// Cannot be used in combination with [startAt].
-  DocumentQuery startAfter(List<dynamic> values) {
-    final jsValues = jsify(values);
-    return new DocumentQuery(nativeInstance.startAfter(jsValues), firestore);
+  DocumentQuery startAfter({DocumentSnapshot snapshot, List<dynamic> values}) {
+    return new DocumentQuery(
+        _wrapPaginatingFunctionCall("startAfter", snapshot, values), firestore);
```
Comment (pulyaevskiy) [inline lib/src/firestore.dart]: Need to update dartdoc to something like this:

```
/// Takes a [snapshot] or a list of [values], creates and returns a new [DocumentQuery]
/// that starts after the provided fields relative to the order of the query.
///
/// Either [snapshot] or [values] can be provided at the same time, not both.
/// The [values] must be in order of [orderBy] filters.

/// Cannot be used in combination with [startAt].
```

Similar changes needed for other paginating methods.

Comment (pulyaevskiy) [reply lib/src/firestore.dart]: nit: Also might be a good idea to add an assert like:

```dart
assert(snapshot == null || values == null, 'Cannot provide both snapshot and values to startAfter.');
```

File: lib/src/firestore.dart [hunk 4]
```diff
@@ -4,6 +4,7 @@
 import 'dart:async';
 import 'dart:js';
 
+import 'package:firebase_admin_interop/js.dart';
```
Comment (pulyaevskiy) [inline lib/src/firestore.dart]: Curious why is this needed here? It only exports `bindings.dart` which is already imported here under `js` prefix.

File: lib/src/firestore.dart [hunk 5]
```diff
@@ -727,4 +789,36 @@ class DocumentQuery {
...
+  /// Specifies the offset of the returned results.
+  DocumentQuery offset(int offset) {
+    assert(offset != null);
+    return new DocumentQuery(nativeInstance.offset(offset), firestore);
+  }
+
+  /// Calls js paginating [method] with [DocumentSnapshot] or List of [values].
+  /// We need to call this method in all paginating methods to fix that Dart
+  /// doesn't support varargs - we need to use [List] to call js function.
+  _wrapPaginatingFunctionCall(
```
Comment (pulyaevskiy) [inline lib/src/firestore.dart]: Needs return type, probably `js.DocumentQuery`.

File: test/firestore_test.dart [hunk 6]
```diff
@@ -63,6 +63,30 @@ void main() {
         expect(nested.getString('author'), 'Isaac Asimov');
       });
 
+      test('get set', () {
```
Comment (pulyaevskiy) [inline test/firestore_test.dart]: This test seems to be duplicating logic from other test cases. Though I understand this is a more lightweight unit test it still feels a bit redundant.

For the `FieldValue` edge cases I'd create separate test cases for each value as integration tests, writing to the database and reading back the results.

Comment (pulyaevskiy) [reply test/firestore_test.dart]: Oh, I see you actually have separate tests for special field values below. I'm a bit confused about the purpose of this test than. :)

File: lib/src/firestore.dart (hunk 2 above)
Comment (alextekartik) [reply lib/src/firestore.dart]: one issue is that the user can read a DocumentData/UpdateData object before sending to firestore (or call toMap/fromMap). If I create a document data and call setServerTimestamp and right call toMap, what would be the value then? And I think calling fromMap(toMap) and toMap(fromMap) should lead to the same object content. Not sure if I'm clear here though...

File: test/firestore_test.dart (hunk 6 above)
Comment (alextekartik) [reply test/firestore_test.dart]: Many setters (setDouble, setBool) did not have any unit test. The goal is to make sure that we can read/get all fields with setter and getter and that the roundTrip fromMap(toMap) was working fine too

File: lib/src/firestore.dart [hunk 7]
```diff
@@ -291,32 +301,41 @@ class _FirestoreData {
...
     setProperty(nativeInstance, key, value);
   }
 
   bool getBool(String key) => (getProperty(nativeInstance, key) as bool);
+
   void setBool(String key, bool value) {
     setProperty(nativeInstance, key, value);
   }
 
+  /// true if the data contains an entry with the given [key]
```
[the remark this answers is not in the document]
Comment (alextekartik) [reply lib/src/firestore.dart]: ok

File: lib/src/firestore.dart (hunk 1 above)
Comment (alextekartik) [reply lib/src/firestore.dart]: If you read the document data before writing it to Firestore, you will get the FieldValue.serverTimestamp (that's how it works on javascript). After you write to firestore, reading the same field contains DateTime object. As I said in my other comment what would be the value if you call fromMap? But I understant your concert as such fields (delete, serverTimestamp) should be write only!

File: lib/src/firestore.dart (hunk 3 above)
Comment (alextekartik) [reply lib/src/firestore.dart]: ok

File: lib/src/firestore.dart (hunk 4 above)
Comment (alextekartik) [reply lib/src/firestore.dart]: ok removed

File: lib/src/firestore.dart (hunk 5 above)
Comment (alextekartik) [reply lib/src/firestore.dart]: ok fixed

File: lib/src/firestore.dart (hunk 2 above)
Comment (alextekartik) [reply lib/src/firestore.dart]: Also if we don't have the FieldValue enum (but yes it could be a class), how could this be used if I want to create a map and call fromMap?

File: lib/src/firestore.dart (hunk 1 above)
Comment (alextekartik) [reply lib/src/firestore.dart]: But yes, I will make it private only (will be use in toMap though)

**Comment by alextekartik** [comment]
I pushed new changes in response to your as always very good suggestions. I've left the extra "get set" test although bit redundant but some setter were not tested and I did not want to mess other tests. I have remove the getFieldValue however the FieldValue type remains so that it can be set either using fromMap or using setFieldValue.

File: lib/src/firestore.dart [hunk 8]
```diff
@@ -22,6 +22,14 @@ js.FieldPath createFieldPath(List<String> fieldNames) {
...
 }
 
+js.FieldValue fieldValueDelete() {
+  return js.admin.firestore.FieldValue.delete();
+}
+
+js.FieldValue fieldValueServerTimestamp() {
+  return js.admin.firestore.FieldValue.serverTimestamp();
+}
+
```
Comment (pulyaevskiy) [inline lib/src/firestore.dart]: These globally available functions bother me quite a bit. I see that you probably followed the pattern I started with `createGeoPoint` and others, so this is my bad.

I thought it would be nice to namespace them under Firestore and realized that we should be able to expose native `FieldValue` field like this:

```dart
/// Represents a Firestore Database and is the entry point for all
/// Firestore operations.
class Firestore {
  /// Sentinel field values that can be used when writing document fields with
  /// `set` or `update`.
  js.FieldValues get fieldValues => js.admin.firestore.FieldValue;

  // the rest of the Firestore class
}
```

So enum and global functions can go away. The `setFieldValue` method can declare it expects native `js.FieldValue` type and there is no need to convert between native values and enum.

WDYT?
If we do it this way I'll probably move other global functions under `Firestore` for consistency (after this PR is merged).

Comment (alextekartik) [reply lib/src/firestore.dart]: you meant fieldValues to be static right (I assume so)? i.e. do you want to access them like this:
```
Firestore.fieldValues.serverTimestamp()
```
or
```
app.firestore().fieldValues.serverTimestamp()
```

> The setFieldValue method can declare it expects native js.FieldValue type

I thought you wanted to avoid exposing javascript types. Ok for me

**Comment by alextekartik** [comment]
I pushed more changes according to your suggestions (assuming Firestore.fieldValues is static)
